sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java

```diff
@@ -815,13 +816,169 @@ public void validate(PipelineOptions options) {
 
       Pipeline p = input.getPipeline();
       final Coder<T> coder = inferCoder(p.getCoderRegistry());
+
       if (getMethod() == Method.BQ_PARALLEL_READ) {
-        return p.apply(org.apache.beam.sdk.io.Read.from(BigQueryParallelReadTableSource.create(
-            getTableProvider(),
-            getRowProtoParseFn(),
-            coder,
-            getBigQueryServices(),
-            getReadSessionOptions())));
+        // When using Method.BQ_PARALLEL_READ to read directly from a table, there are no temporary
+        // resources to clean up; the Read transform can be returned directly.
+        if (table != null) {
+          return p.apply(org.apache.beam.sdk.io.Read.from(BigQueryParallelReadTableSource.create(
+              getTableProvider(),
+              getRowProtoParseFn(),
+              coder,
+              getBigQueryServices(),
+              getReadSessionOptions())));
+        }
+
+        // When using Method.BQ_PARALLEL_READ to read the results of a query, the underlying
+        // dataset and table must remain live while the data is being read and then be cleaned up
+        // afterwards.
+        PCollection<String> jobIdTokenCollection =
+            p.apply("TriggerJobIdCreation", Create.of("ignored"))
+                .apply("CreateJobId",
+                    MapElements.via(
+                        new SimpleFunction<String, String>() {
+                          @Override
+                          public String apply(String input) {
+                            return BigQueryHelpers.randomUUIDString();
+                          }
+                        }));
+
+        final TupleTag<ParallelRead.Session> sessionTag = new TupleTag<>();
+        final TupleTag<ParallelRead.ReadLocation> readLocationTag = new TupleTag<>();
+        final TupleTag<String> tableReferenceTag = new TupleTag<>();
+
+        PCollectionTuple tuple =
+            jobIdTokenCollection.apply(
+                "ExecuteQuery",
```

Comment:
This one is not executing the query though, it just returns ReadLocation's - does that mean it splits the query?

Reply:
Under the covers, what's actually happening here is that we're:
- Executing a dry run of the query in order to determine the appropriate location for the temporary dataset and table to receive the results
- Creating the temporary dataset and table
- Executing the actual query
- Creating a read session against the temporary table using the read API

Splitting only happens in that last step (a single read session is split into many streams). I can refactor this code a bit in order to make this clearer.